packages/next/src/client/app-index.tsx

```diff
@@ -172,9 +171,36 @@ const initialServerResponse = createFromReadableStream(readable, {
 })
 
 function ServerRoot(): React.ReactNode {
-  const initialResponse = use<InitialRSCPayload>(initialServerResponse)
+  const initialRSCPayload = use<InitialRSCPayload>(initialServerResponse)
+  const initialState = createInitialRouterState({
+    buildId: initialRSCPayload.b,
+    initialFlightData: initialRSCPayload.f,
+    initialCanonicalUrl: initialRSCPayload.c,
+    initialParallelRoutes: new Map(),
+    location: window.location,
+    couldBeIntercepted: initialRSCPayload.i,
+  })
+
+  const actionQueue = createMutableActionQueue(initialState)
```

Comment:
I'd lift these up out of the render pathway. If we refactored this in the future to have a parent and that parent updated these would get recreated which is probably bad.

probably need something like

```
const pendingActionQueue = initialServerResponse.then(r => createMutableActionQueue(createInitialRouterState(...)))

function ServerRoot() {
  const initialRSCPayload = use<InitialRSCPayload>(initialServerResponse)
  const actionQueue = use(pendingActionQueue)
  ...
}
```